mesonbuild/modules/python.py

```diff
@@ -0,0 +1,432 @@
+# Copyright 2018 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import json
+
+from pathlib import Path
+from .. import mesonlib
+from . import ExtensionModule
+from mesonbuild.modules import ModuleReturnValue
+from . import permittedSnippetKwargs
+from ..interpreterbase import (
+    noPosargs, noKwargs, permittedKwargs,
+    InterpreterObject, InvalidArguments
+)
+from ..interpreter import shlib_kwargs, ExternalProgramHolder
+from .. import mlog
+from ..environment import detect_cpu_family
+from ..dependencies.base import (
+    DependencyMethods, ExternalDependency,
+    ExternalProgram, PkgConfigDependency,
+    NonExistingExternalProgram
+)
+
+mod_kwargs = set(['subdir'])
+mod_kwargs.update(shlib_kwargs)
+mod_kwargs -= set(['install_dir', 'install_rpath'])
+
+
+def run_command(python, command):
+    _, stdout, _ = mesonlib.Popen_safe(python.get_command() + [
+        '-c',
+        command])
+
+    return stdout.strip()
+
+
+class PythonDependency(ExternalDependency):
+    def __init__(self, python_holder, environment, kwargs):
+        super().__init__('python', environment, None, kwargs)
+        self.name = 'python'
+        self.static = kwargs.get('static', False)
+        self.version = python_holder.version
+        self.platform = python_holder.platform
+        self.pkgdep = None
+        self.variables = python_holder.variables
+        self.paths = python_holder.paths
+        if mesonlib.version_compare(self.version, '>= 3.0'):
+            self.major_version = 3
+        else:
+            self.major_version = 2
+
+        if DependencyMethods.PKGCONFIG in self.methods:
+            pkg_version = self.variables.get('LDVERSION') or self.version
+            pkg_path = self.variables.get('LIBPC')
+            old_pkg_path = os.environ.get('PKG_CONFIG_PATH')
+
+            if pkg_path:
+                os.environ['PKG_CONFIG_PATH'] = pkg_path
```

Comment:
`os.environ` has to be exception safe. Also this looks a bit fragile, because it depends on `PkgConfigDependency` not doing anything itself with dependencies.

Also pkg-config should not search in it's hardcoded pathes. Maybe setting `PKG_CONFIG_LIBDIR` to an empty string might work.

NB: It's really hard to see changes when you amend new changes into the old commits. Usually it is better to create temporary commits for the reviewers and later squash them after review 'stages'. (Sadly github does not have a gui way to rebase --autosquash, so manual squashing is needed)

Reply:
> os.environ has to be exception safe. 

Yes, it is restored when an exception is caught

> Also this looks a bit fragile, because it depends on PkgConfigDependency not doing anything itself with dependencies.

seems robust enough in python's case to me

> Also pkg-config should not search in it's hardcoded pathes. Maybe setting PKG_CONFIG_LIBDIR to an empty string might work.

Or just use `PKG_CONFIG_LIBDIR` instead of `PKG_CONFIG_PATH`?

Reply:
I didn't notice that the restore code is there twice: Once for the success case and once after possibly catching all exceptions. So while it's a bit duplicated it should work. I think one finally block would be nicer (or even an context manager)

With `PKG_CONFIG_LIBDIR` i think `PKG_CONFIG_PATH` would need to be unset. Not sure which way around is better. But `PKG_CONFIG_PATH` unset is at least a known working case.

Reply:
I'll refactor that tomorrow, as I said it was mostly taken from meson, including the early return which makes things a bit awkward

Reply:
k, tomorrow ended up meaning 11 days later, sorry :)